### ISSUE TITLE
Hangs when select and edit in a 15MB file.

### ISSUE THREAD

**Issue by grv07** [helix-editor/helix#7396]
### Summary

The file is 14.9 MB in size sample data
```
Ok("any-string")
Ok("any-string")
Ok("any-string")
... 388892 lines like that
```

I try "% + s + Ok\(" and it hangs

### Reproduction Steps

 

I tried this:

1. `hx 15mb_file_name`
2. % + s + Ok\( + enter 

I expected this to happen:
I should be able to edit the text.

Instead, this happened:
It hangs and does not respond.

### Helix log

Unable to get logs since it is not responding at all.

### Platform

Linux (Ubuntu 23.04)

### Terminal Emulator

alacrity + zellij

### Helix Version

master

![image](https://github.com/helix-editor/helix/assets/11697220/04a089cc-7171-4787-8f8d-802fc39832ca)

**Comment by kirawi** [comment]
I'm very certain there's already an issue for this, but I can't seem to find it through a quick search. Regardless, this is just a limitation of having a lot of active selections at once.

**Comment by CptPotato** [comment]
I've worked with very large files (hundreds of Mb) and even if helix gets slow it does respond after a little while.
Have you tried escaping the parenthesis `Ok\(`? If I remember correctly more complex expressions get very slow very quickly and tend to run out of memory on such large files. I'm having no issues with 1000000 lines of `Ok(hi)`.

Also, one way to speed this up is to yank the search text first: `"/y`
Then helix has to match the regex just once, and not go over the whole file for every keystroke.

**Comment by pascalkuthe** [comment]
Poeple seem to have missread the issue description: The problem is not the regex selection that is pretty fast but actually editing once there is a selection is insanely slow. I whipped up a quick flamegraph and now know why: Almost all time is spend updating the selection (ropey is so fast it doesn't even show up). The reason is simple: muiticursos edits currently have a time complexity `O(N^2)` for the number of cursors. Fixing that will  be a bit involved but should be possible

**Comment by the-mikedavis** [comment]
Fixed in #7408
